File: airesources/Java/hlt/Ship.java
package hlt;

public class Ship extends Entity {

    public enum DockingStatus { Undocked, Docking, Docked, Undocking }

    private final DockingStatus dockingStatus;
    private final long dockedPlanet;
    private final short dockingProgress;
    private final short weaponCooldown;

    public Ship(final short owner, final long id, final double xPos, final double yPos,
                final short health, final DockingStatus dockingStatus, final long dockedPlanet,
                final short dockingProgress, final short weaponCooldown) {

        super(owner, id, xPos, yPos, health, Constants.SHIP_RADIUS);

        this.dockingStatus = dockingStatus;
        this.dockedPlanet = dockedPlanet;
        this.dockingProgress = dockingProgress;
        this.weaponCooldown = weaponCooldown;
    }

    public short getWeaponCooldown() {
        return weaponCooldown;
    }

    public DockingStatus getDockingStatus() {
        return dockingStatus;
    }

    public short getDockingProgress() {
        return dockingProgress;
    }

    public long getDockedPlanet() {
        return dockedPlanet;
    }

    public boolean canDock(final Planet planet) {
        return getDistanceTo(planet) <= Constants.DOCK_RADIUS + planet.getRadius();
    }
}

File: airesources/Scala/hlt/Ship.scala
package hlt

object Ship {

  def values: List[DockingStatus] = List(Undocked, Docking, Docked, Undocking)

  sealed trait DockingStatus

  case object Undocked extends DockingStatus

  case object Docking extends DockingStatus

  case object Docked extends DockingStatus

  case object Undocking extends DockingStatus

}

class Ship(override val owner: Option[Short],
           override val id: Long,
           override val xPos: Double,
           override val yPos: Double,
           override val health: Short,
           val dockingStatus: Ship.DockingStatus,
           val dockedPlanet: Long,
           val dockingProgress: Short,
           val weaponCooldown: Short)
    extends Entity(owner, id, xPos, yPos, health, Constants.SHIP_RADIUS) {

  def canDock(planet: Planet): Boolean =
    getDistanceTo(planet) <= Constants.DOCK_RADIUS + planet.radius

  override def toString: String =
    "Ship[" + super.toString + ", dockingStatus=" + dockingStatus + ", dockedPlanet=" + dockedPlanet +
      ", dockingProgress=" + dockingProgress + ", weaponCooldown=" + weaponCooldown + "]"
}
